File: backend-laravel/app/Models/Rol.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\SoftDeletes;

/**
 * Rol libre, definido por la institución (tabla `roles`). El nombre lo
 * elige la institución (preceptor, tutor, celador...); lo que puede hacer
 * cada rol surge exclusivamente de los permisos fijos que se le asignan
 * vía permisos().
 *
 * SoftDeletes (deleted_at): igual que en Usuario, `uq_roles_nombre` no
 * incluye deleted_at — reusar el nombre de un rol borrado exige
 * restaurarlo (withTrashed()) en vez de crear uno nuevo con el mismo
 * nombre.
 */
class Rol extends Model
{
    use SoftDeletes;

    protected $table = 'roles';

    protected $primaryKey = 'id_rol';

    protected $fillable = [
        'nombre',
        'descripcion',
        'activo',
    ];

    protected function casts(): array
    {
        return [
            'activo' => 'boolean',
            'deleted_at' => 'datetime',
        ];
    }

    public function permisos(): BelongsToMany
    {
        return $this->belongsToMany(
            Permiso::class,
            'roles_permisos',
            'rol_id',
            'permiso_id',
            'id_rol',
            'id_permiso'
        );
    }

    public function usuarios(): BelongsToMany
    {
        return $this->belongsToMany(
            Usuario::class,
            'usuarios_roles',
            'rol_id',
            'usuario_id',
            'id_rol',
            'id_usuario'
        );
    }
}

File: app_flutter/lib/models/rol.dart
import 'permiso.dart';

/// Espejo de lo que devuelve `RolesController` (backend-laravel) — un
/// rol libre, definido por la institución (ej. "preceptor",
/// "director"). Lo que puede hacer surge exclusivamente de `permisos`
/// (subconjunto del catálogo fijo, ver `App\Models\Rol`).
class Rol {
  const Rol({
    required this.id,
    required this.nombre,
    required this.descripcion,
    required this.activo,
    required this.cantidadUsuarios,
    required this.permisos,
  });

  factory Rol.fromJson(Map<String, dynamic> json) {
    final permisosJson = json['permisos'] as List<dynamic>? ?? const [];
    return Rol(
      id: json['id_rol'] as int,
      nombre: json['nombre'] as String,
      descripcion: json['descripcion'] as String?,
      activo: json['activo'] as bool,
      cantidadUsuarios: json['cantidad_usuarios'] as int? ?? 0,
      permisos: permisosJson
          .map((item) => Permiso.fromJson(item as Map<String, dynamic>))
          .toList(growable: false),
    );
  }

  final int id;
  final String nombre;
  final String? descripcion;
  final bool activo;
  final int cantidadUsuarios;
  final List<Permiso> permisos;
}
